### PR TITLE
Offline data enhancements

### PR DESCRIPTION
**Pull request by nCastle1**
These changes make the use of offline data more clear and enhance the separation between samples’ conceptual demonstration and the mechanics of the sample viewer.

### CODE REVIEW

**Comment by nCastle1** [comment]
@macca400 Can you please review these changes? These are necessary to support the changes to the doc build process (samples solution generator). The changes include:

- Separating out offline data functionality with region tags
- Adding a `DataFileNames` property to the metadata.json files; in addition to DataItemIds, this metadata identifies which specific files are going to be used.

This needs to be merged before any changes are merged in for the sample solution generator.

File: src/Android/Xamarin.Android/Samples/Layers/FeatureLayerDictionaryRenderer/FeatureLayerDictionaryRenderer.cs [hunk 1]
```diff
@@ -111,23 +101,46 @@ private void CreateLayout()
...
+            string folder = DataManager.GetDataFolder();
+
+            // Get the full path
+            string filepath = Path.Combine(folder, "SampleData", "FeatureLayerDictionaryRenderer", "mil2525d.stylx");
+
+            // Check if the file exists
+            if (!File.Exists(filepath))
+            {
+                // Download the file
+                await DataManager.GetData("e34835bf5ec5430da7cf16bb8c0b075c", "FeatureLayerDictionaryRenderer");
```
[the remark this answers is not in the document]
Comment (macca400) [reply src/Android/Xamarin.Android/Samples/Layers/FeatureLayerDictionaryRenderer/FeatureLayerDictionaryRenderer.cs]: Check out https://github.com/Esri/arcgis-runtime-samples-dotnet/blob/461277f1076fd94c8d021418da375f65a5939404/src/UWP/ArcGISRuntime.UWP.Viewer/SamplePage.xaml.cs

specifically 
```
if(SampleManager.Current.SelectedSample.RequiresOfflineData == true)
            {
                foreach (var id in SampleManager.Current.SelectedSample.DataItemIds)
                {
                    DataManager.GetData(id, SampleManager.Current.SelectedSample.Name);
                }
            }
```

[the remark this answers is not in the document]
Comment (macca400) [reply src/Android/Xamarin.Android/Samples/Layers/FeatureLayerDictionaryRenderer/FeatureLayerDictionaryRenderer.cs]: https://github.com/Esri/arcgis-runtime-samples-dotnet/blob/OfflineMapTask/src/WPF/ArcGISRuntime.WPF.Samples/Samples/Offline/OpenMobileMapPackage/OpenMobileMapPackage.xaml.cs

[the remark this answers is not in the document]
Comment (nCastle1) [reply src/Android/Xamarin.Android/Samples/Layers/FeatureLayerDictionaryRenderer/FeatureLayerDictionaryRenderer.cs]: I found that in some cases on WPF, the user could end up in a sample with offline data, decide not to click the 'download offline data' link, and crash the sample viewer. As a result, the check-and-download is needed for that platform, and I wanted to keep the code consistent. Now with how the sample solution generator is implemented, the item ID needs to be between the region tag regardless, even if that is just a comment. 

I think I'll go through and remove the check and just leave the item IDs in a comment.

**Comment by nCastle1** [comment]
@macca400 Because of how the sample viewer is implemented on WPF & UWP, the suggested change can't be made; it seems that the sample viewer attempts to load the sample even before the user has an opportunity to download offline data. I'll be making the changes to the other platforms as possible.

**Comment by nCastle1** [comment]
Closing this as all code review changes were made (as applicable).

**Comment by macca400** [comment]
@nCastle1 I might have just forced WPF to download rather than giving the option to click the download link. I forgot it was forwarding to a download page  

Swapping the following code in MainWindow.xaml.cs

```
          if (SampleManager.Current.SelectedSample.RequiresOfflineData == true)
                {
                    var sampleDataPath = Path.Combine(DataManager.GetDataFolder(), "SampleData", SampleManager.Current.SelectedSample.Name);
                    if (!Directory.Exists(sampleDataPath))
                        {
                        data.IsChecked = true;
                        SampleContainer.Visibility = Visibility.Collapsed;
                        DescriptionContainer.Visibility = Visibility.Collapsed;
                        DataContainer.Visibility = Visibility.Visible;
                        }               
                }
```
for something more like... (so data has to be downloaded if the sample is selected)  
```
                if (item.RequiresOfflineData)
                {
                    foreach (string id in item.DataItemIds)
                    { 
                        await DataManager.GetData(id, item.SampleName);
                    }
                }
```
Probably not in the scope of this issue. But i'd be keen to rip out the UI download link stuff we currently have in WPF. Moving to an approach where if the sample is selected the data is downloaded. Else where on the home screen add a Button download all required data in one go.

**Pull request merged**